clash-prelude/src/Clash/Sized/Internal/Index.hs

```diff
@@ -133,10 +138,17 @@ import Clash.XException
 -- >>> 2 * 4 :: Index 8
 -- *** Exception: X: Clash.Sized.Index: result 8 is out of bounds: [0..7]
 -- ...
+#if MIN_VERSION_base(4,15,0)
+data Index (n :: Nat) =
+    -- | The constructor, 'I', and the field, 'unsafeToInteger', are not
+    -- synthesizable.
+    I { unsafeToInteger :: !Integer }
+#else
 newtype Index (n :: Nat) =
     -- | The constructor, 'I', and the field, 'unsafeToInteger', are not
     -- synthesizable.
     I { unsafeToInteger :: Integer }
+#endif
```

Comment:
This changes the ability to do `coerce` on `Index`es. Wouldn't it make sense to have the same behavior for all ghc versions? Did this change show up in benchmarks at all?

Reply:
No, see my earlier comment, the ghc-9 change makes performance likely worse; but it's the only work-around that I know of.